compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/ReflectiveAccess.kt

```diff
@@ -80,14 +83,15 @@ internal class ReflectiveAccessLowering(
     // Fragments are transformed in a post-order traversal: children first,
     // then parent. This obscures, in particular, dispatch receivers, that go
     // from `IrGetObjectValue` calls to blocks implementing the corresponding
-    // reflective access . We record these _before_ transformation, in order to
+    // reflective access. We record these _before_ transformation, in order to
     // later predict the compilation strategy for fields. See the uses of
     // `fieldLocationAndReceiver`.
     val callsOnCompanionObjects: MutableMap<IrCall, IrClassSymbol> = mutableMapOf()
 
+    @OptIn(ObsoleteDescriptorBasedAPI::class)
     private fun recordCompanionObjectAsDispatchReceiver(expression: IrCall) {
-        if ((expression.dispatchReceiver as? IrGetObjectValue)?.symbol?.owner?.isCompanion == true) {
-            callsOnCompanionObjects[expression] = (expression.dispatchReceiver!! as IrGetObjectValue).symbol
+        if (((expression.dispatchReceiver as? IrGetField)?.symbol?.owner?.type as? IrSimpleTypeImpl)?.classifier?.descriptor?.isCompanionObject() == true) {
```

Comment:
Is there a better way to detect calls on companion objects? This feels extremely ad-hoc, but works.

Reply:
Seems fine, but let's avoid using descriptors, and computing same things multiple times:
```
    private fun recordCompanionObjectAsDispatchReceiver(expression: IrCall) {
        val dispatchReceiver = expression.dispatchReceiver as? IrGetField ?: return
        val dispatchReceiverType = dispatchReceiver.symbol.owner.type as? IrSimpleType ?: return
        val klass = dispatchReceiverType.classOrNull
        if (klass != null && klass.owner.isCompanion) {
            callsOnCompanionObjects[expression] = klass
        }
    }
```

Reply:
Agreed! Done.